crates/sparrow-scheduler/src/queue.rs

```diff
@@ -0,0 +1,148 @@
+//! Provide the global and per-thread (local) work queues.
+//!
+//! These wrap the currently used crate to make it easy to swap in
+//! different implementations.
+
+pub trait Queue<T> {
+    /// Push an item to the *front* of the local queue.
+    ///
+    /// The item will be placed in the LIFO slot (next thing this thread runs).
+    /// If there was already an item in that slot, it will be displaced.
+    ///
+    /// If the local queue is full, it will move half of its items to the global queue.
+    ///
+    /// When called on a global queue, items will be pushed to that.
+    fn push(&mut self, item: T);
+
+    /// Push an item to the *end* of the local queue, skipping the LIFO slot.
+    ///
+    /// This can be used to give other tasks a chance to run. Otherwise, there’s
+    /// a risk that one task will completely take over a thread in a push-pop
+    /// cycle due to the LIFO slot.
+    ///
+    /// When called on a global queue, items will be pushed to that.
+    fn push_yield(&mut self, item: T);
+
+    /// Push an item onto the global queue.
+    fn push_global(&self, item: T);
```

Comment:
What is the difference between this function and the one preceding it?

The last comment on the preceding function spells out how it would work for a global queue.

Reply:
When called on a local queue, `push_global` will push to the global queue, while `push_yield` will push to the end of the local queue.

Reply:
If the queue is global isn't easier to make that function part of the global queue and leave it out of this trait? Or are you looking at having the same trait for both global and local queues?

Reply:
Trying to have the same trait. The reason being that it lets us call `add_input(...)` from outside of a worker thread (eg., from tokio) and pass the global queue (so we default to queuing work) and `add_input(...)` from the worker thread adds to the local queue.

On the other hand, I could be pretty easily convinced to treat sources as special, which would let us use a different trait for them, that received the global queue only.